components/tidb_qe/src/executor/run.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Arc;
+
+use kvproto::coprocessor::KeyRange;
+use protobuf::Message;
+use tipb::executor::{self, ExecType, ExecutorExecutionSummary};
+use tipb::select::{Chunk, DAGRequest, SelectResponse, StreamResponse};
+
+use tikv_util::deadline::Deadline;
+
+use super::Executor;
+use crate::execute_stats::*;
+use crate::expr::EvalConfig;
+use crate::metrics::*;
+use crate::storage::{IntervalRange, Storage};
+use crate::Result;
+
+pub struct ExecutorsRunner<SS> {
+    deadline: Deadline,
+    executor: Box<dyn Executor<StorageStats = SS> + Send>,
+    output_offsets: Vec<u32>,
+    batch_row_limit: usize,
+    collect_exec_summary: bool,
+    exec_stats: ExecuteStats,
+}
+
+/// Builds a normal executor pipeline.
+///
+/// Normal executors iterate rows one by one.
+pub fn build_executors<S: Storage + 'static, C: ExecSummaryCollector + 'static>(
+    exec_descriptors: Vec<executor::Executor>,
+    storage: S,
+    ranges: Vec<KeyRange>,
+    ctx: Arc<EvalConfig>,
+    is_streaming: bool,
+) -> Result<Box<dyn Executor<StorageStats = S::Statistics> + Send>> {
+    let mut exec_descriptors = exec_descriptors.into_iter();
+    let first = exec_descriptors
+        .next()
+        .ok_or_else(|| unknown_err!("No executor specified"))?;
+
+    let mut src = build_first_executor::<_, C>(first, storage, ranges, is_streaming)?;
+    let mut summary_slot_index = 0;
+
+    for mut exec in exec_descriptors {
+        summary_slot_index += 1;
+
+        let curr: Box<dyn Executor<StorageStats = S::Statistics> + Send> = match exec.get_tp() {
+            ExecType::TypeTableScan | ExecType::TypeIndexScan => {
```

Comment:
It's better to keep consistent with the batch version `build_executors` (use the `_` match arm to emit this error).